runtime/src/local_runqueue_mts.c

```diff
@@ -0,0 +1,281 @@
+#include <stdint.h>
+
+#include "arch/context.h"
+#include "client_socket.h"
+#include "current_sandbox.h"
+#include "debuglog.h"
+#include "global_request_scheduler.h"
+#include "local_runqueue.h"
+#include "local_runqueue_mts.h"
+#include "panic.h"
+#include "priority_queue.h"
+#include "sandbox_functions.h"
+#include "runtime.h"
+
+__thread static struct priority_queue *local_runqueue_mts_guaranteed;
+__thread static struct priority_queue *local_runqueue_mts_default;
+
+extern __thread struct priority_queue *worker_thread_timeout_queue;
+
+/**
+ * Get Per-Worker-Module priority for Priority Queue ordering
+ * @param element the PWM
+ * @returns the priority of the head of the PWM
+ */
+static inline uint64_t
+perworker_module_get_priority(void *element)
+{
+	struct perworker_module_sandbox_queue *pwm     = (struct perworker_module_sandbox_queue *)element;
+	struct sandbox *                       sandbox = NULL;
+	priority_queue_top_nolock(pwm->sandboxes, (void **)&sandbox);
+	return (sandbox) ? sandbox->absolute_deadline : UINT64_MAX;
+}
+
+/**
+ * Checks if the run queue is empty
+ * @returns true if empty. false otherwise
+ */
+bool
+local_runqueue_mts_is_empty()
+{
+	return priority_queue_length_nolock(local_runqueue_mts_guaranteed) == 0
+	       && priority_queue_length_nolock(local_runqueue_mts_default) == 0;
+}
+
+/**
+ * Adds a sandbox to the run queue
+ * @param sandbox
+ * @returns pointer to sandbox added
+ */
+void
+local_runqueue_mts_add(struct sandbox *sandbox)
+{
+	assert(sandbox != NULL);
+
+	struct perworker_module_sandbox_queue *pwm               = &sandbox->module->pwm_sandboxes[worker_thread_idx];
+	struct priority_queue *                destination_queue = local_runqueue_mts_default;
+
+	uint64_t prev_pwm_deadline = priority_queue_peek(pwm->sandboxes);
+
+	/* Add the sandbox to the per-worker-module (pwm) queue */
+	int rc = priority_queue_enqueue_nolock(pwm->sandboxes, sandbox);
+	if (rc == -ENOSPC) panic("Per Worker Module queue is full!\n");
+	// debuglog("Added a sandbox to the PWM of module '%s'", sandbox->module->name);
+
+	if (priority_queue_length_nolock(pwm->sandboxes) == 1) {
+		pwm->mt_class = MT_DEFAULT;
+		if (sandbox->module->remaining_budget > 0) {
+			pwm->mt_class     = MT_GUARANTEED;
+			destination_queue = local_runqueue_mts_guaranteed;
+		}
+
+		/* Add sandbox module to the worker's timeout queue if guaranteed tenant and only if first sandbox*/
+		if (sandbox->module->replenishment_period > 0) {
+			struct module_timeout *mt = malloc(sizeof(struct module_timeout));
+
+			mt->timeout = get_next_timeout_of_module(sandbox->module->replenishment_period, __getcycles());
+			mt->module  = sandbox->module;
+			priority_queue_enqueue_nolock(worker_thread_timeout_queue, mt);
+		}
+
+		rc = priority_queue_enqueue_nolock(destination_queue, pwm);
+
+		return;
+	}
+
+	if (sandbox->absolute_deadline < prev_pwm_deadline) {
```

Comment:
This never runs if the modules sandbox queue is length 1. I can't tell if that is OK or not. I *might* expect a check on the absolute deadline in the previous block of code. Not sure.

Reply:
When the queue length is one, there is no need to delete & add the module, just add (line 81).